x/gamm/twap/store_test.go

```diff
@@ -0,0 +1,181 @@
+	tPlusOne := baseTime.Add(time.Second)
+	baseRecord := newEmptyPriceRecord(1, baseTime, "tokenB", "tokenA")
+	tPlusOneRecord := newEmptyPriceRecord(1, tPlusOne, "tokenB", "tokenA")
+	tests := map[string]struct {
+		recordsToSet    []types.TwapRecord
+		poolId          uint64
+		expectedRecords []types.TwapRecord
+	}{
+		"set single record": {
+			[]types.TwapRecord{baseRecord},
+			1,
+			[]types.TwapRecord{baseRecord},
+		},
+		"query non-existent pool": {
+			[]types.TwapRecord{baseRecord},
+			2,
+			[]types.TwapRecord{},
+		},
+		"set two records": {
+			[]types.TwapRecord{baseRecord, tPlusOneRecord},
+			1,
+			[]types.TwapRecord{tPlusOneRecord},
+		},
+		"set two records, reverse order": {
+			// The last record, independent of time, takes precedence for most recent.
+			[]types.TwapRecord{tPlusOneRecord, baseRecord},
+			1,
+			[]types.TwapRecord{baseRecord},
+		},
+		"set multi-asset pool record": {
+			[]types.TwapRecord{
+				newEmptyPriceRecord(1, baseTime, "tokenB", "tokenA"),
+				newEmptyPriceRecord(1, baseTime, "tokenC", "tokenB"),
+				newEmptyPriceRecord(1, baseTime, "tokenC", "tokenA")},
+			1,
+			[]types.TwapRecord{
+				newEmptyPriceRecord(1, baseTime, "tokenB", "tokenA"),
+				newEmptyPriceRecord(1, baseTime, "tokenC", "tokenA"),
+				newEmptyPriceRecord(1, baseTime, "tokenC", "tokenB")},
+		},
+	}
+
+	for name, test := range tests {
+		s.Run(name, func() {
+			s.SetupTest()
+			for _, record := range test.recordsToSet {
+				s.twapkeeper.StoreNewRecord(s.Ctx, record)
+			}
+			actualRecords, err := s.twapkeeper.GetAllMostRecentRecordsForPool(s.Ctx, test.poolId)
+			s.Require().NoError(err)
+			s.Require().Equal(test.expectedRecords, actualRecords)
+		})
+	}
+}
+
+// TestGetAllMostRecentRecordsForPool takes a list of records as test cases,
+// and runs storeNewRecord for everything in sequence.
+// Then it runs GetRecordAtOrBeforeTime, and sees if its equal to expected
+func (s *TestSuite) TestGetRecordAtOrBeforeTime() {
+	type getRecordInput struct {
+		poolId      uint64
+		t           time.Time
+		asset0Denom string
+		asset1Denom string
+	}
+	defaultInputAt := func(t time.Time) getRecordInput { return getRecordInput{1, t, "tokenB", "tokenA"} }
+	wrongPoolIdInputAt := func(t time.Time) getRecordInput { return getRecordInput{2, t, "tokenB", "tokenA"} }
+	defaultRevInputAt := func(t time.Time) getRecordInput { return getRecordInput{1, t, "tokenA", "tokenB"} }
+	baseRecord := newEmptyPriceRecord(1, baseTime, "tokenB", "tokenA")
+	tMin1 := baseTime.Add(-time.Second)
+	tMin1Record := newEmptyPriceRecord(1, tMin1, "tokenB", "tokenA")
+	tPlus1 := baseTime.Add(time.Second)
+	tPlus1Record := newEmptyPriceRecord(1, tPlus1, "tokenB", "tokenA")
+
+	tests := map[string]struct {
+		recordsToSet   []types.TwapRecord
+		input          getRecordInput
+		expectedRecord types.TwapRecord
+		expErr         bool
+	}{
+		"no entries":            {[]types.TwapRecord{}, defaultInputAt(baseTime), baseRecord, true},
```

Comment:
please consider making these named and each test case field on a new line for readability

Reply:
I found this more readable personally